## Data Cleaning and Analysis for used cars from eBay.

In this project, we will clean and analysis the Data set of used car from eBayKleinanzeigen, a classifieds section of the German eBay website.

The dataset was originally scraped and uploaded to kaggle.

The data dictionary provided with data is as follows:

- dateCrawled - When this ad was first crawled. All field-values are taken from this date.
- name - Name of the car.
- seller - Whether the seller is private or a dealer.
- offerType - The type of listing
- price - The price on the ad to sell the car.
- abtest - Whether the listing is included in an A/B test.
- vehicleType - The vehicle Type.
- yearOfRegistration - The year in which the car was first registered.
- gearbox - The transmission type.
- powerPS - The power of the car in PS.
- model - The car model name.
- kilometer - How many kilometers the car has driven.
- monthOfRegistration - The month in which the car was first - registered.
- fuelType - What type of fuel the car uses.
- brand - The brand of the car.
- notRepairedDamage - If the car has a damage which is not yet repaired.
- dateCreated - The date on which the eBay listing was created.
- nrOfPictures - The number of pictures in the ad.
- postalCode - The postal code for the location of the vehicle.
- lastSeenOnline - When the crawler saw this ad last online.

In [1]:
# Let's import the libraries.
import pandas as pd
import numpy as np

# Reading Data set.

autos = pd.read_csv('autos.csv', encoding="Latin-1")


In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

## Lets know more about the Dataset.

In [3]:
# Getting infomtion about the dtypes.
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

In [4]:
# Below are the columns with dtypes 'int64'

autos.dtypes[autos.dtypes == 'int64']

yearOfRegistration     int64
powerPS                int64
monthOfRegistration    int64
nrOfPictures           int64
postalCode             int64
dtype: object

In [5]:
# Below are the columns with dtypes 'object'

autos.dtypes[autos.dtypes != 'int64']

dateCrawled          object
name                 object
seller               object
offerType            object
price                object
abtest               object
vehicleType          object
gearbox              object
model                object
odometer             object
fuelType             object
brand                object
notRepairedDamage    object
dateCreated          object
lastSeen             object
dtype: object

In [6]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

From the above few rows of dataset we can clearly see that, columns like 'price', 'odometer' represent the numerical values.
But they are string type and need to change to dtypes "int". 

There is not consistancy in the column name. With similar text pattern and meaningful column name, data analysis will be much easier.

Some spelling mistake need to correct in columns like 'seller','greabox' and need to start columns with capital letter.

## Cleaning Column Names

Here, we will change following columns name from camelcase to snakecase:

- yearOfRegistration to registration_year

- monthOfRegistration to registration_month

- notRepairedDamage to unrepaired_damage

- dateCreated to ad_created


In [7]:
# Let's create the function that helps us changing column name.

def change_col_name(old_col_name, new_col_name):
    autos.rename({old_col_name : new_col_name}, axis=1, inplace=True)

In [8]:
data = {
    'yearOfRegistration' : 'registration_year',
    'monthOfRegistration' : 'registration_month',
    'notRepairedDamage' : 'unrepaired_damage',
    'dateCreated' : 'ad_created'
    }

for key,value in data.items():
    change_col_name(key, value)
    

In [9]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuelType', 'brand',
       'unrepaired_damage', 'ad_created', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [10]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  registration_year  \
0  privat   Angebot  $5,000  control         bus               2004   
1  privat   Angebot  $8,500  control   limousine               1997   
2  privat   Angebot  $8,990     test   limousine               2009   
3  privat   Angebot  $4,350  control  kleinwagen               2007   
4  privat   Angebot  $1,350     test       kombi               2003   

     gearbox  powerPS   model   odometer  registration_month fuelType  \
0    manuell      158  andere  150,000km                   3      lpg   
1  automatik      286     7er  150,000km                   6   benzin   
2    manuell      102    golf   70,000km                   7   benzin   
3  automatik       71  fortwo   70,000km                   6   benzin   
4    manuell        0   focus  150,000km                   7   benzin   

        brand unrepaired_damage           ad_created  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

## Initial Exploration and Cleaning.

In [11]:
autos.describe(include='all')

dateCrawled         name  seller offerType  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-10 15:36:24  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicleType  registration_year  gearbox       powerPS  model  \
count        44905       50000.000000    47320  50000.000000  47242   
unique           8                NaN        2           NaN    245   
top      limousine                NaN  manuell           NaN   golf   
freq         12859                NaN    36993           NaN   4024   
mean           NaN        2005.073280      NaN    116.355920    NaN   
std            NaN         105.712813      NaN    209.216627    NaN   
min            NaN        1000.000000      NaN      0.000000    NaN   
25%            NaN        1999.000000      NaN     70.000000    NaN   
50%            NaN        2003.000000      NaN    105.000000    NaN   
75%            NaN        2008.000000      NaN    150.000000    NaN   
max            NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuelType       brand unrepaired_damage  \
count       50000        50000.000000    45518       50000             40171   
unique         13                 NaN        7          40                 2   
top     150,000km                 NaN   benzin  volkswagen              nein   
freq        32424                 NaN    30107       10687             35232   
mean          NaN            5.723360      NaN         NaN               NaN   
std           NaN            3.711984      NaN         NaN               NaN   
min           NaN            0.000000      NaN         NaN               NaN   
25%           NaN            3.000000      NaN         NaN               NaN   
50%           NaN            6.000000      NaN         NaN               NaN   
75%           NaN            9.000000      NaN         NaN               NaN   
max           NaN           12.000000      NaN         NaN               NaN   

                 ad_created  nrOfPictures    postalCode             lastSeen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

In [12]:
# In below 'price' and 'odometer' columns. Numeric values is stored as a text. 
# we need to change it to dtypes as 'int'.
autos[['price','odometer']].head()

price   odometer
0  $5,000  150,000km
1  $8,500  150,000km
2  $8,990   70,000km
3  $4,350   70,000km
4  $1,350  150,000km

In [13]:
# Correcting data and changing dtypes for column 'price'.
price = autos['price']
price_without_dollar = []

for each in price:
    if each.startswith('$'):
        result = each.replace('$', "")
        result = result.replace(',', "")
        price_without_dollar.append(result)
    else:
        each = each.replace(",", "")
        price_without_dollar.append(each)
        
autos['price'] = price_without_dollar
autos['price'] = autos['price'].astype(int)

In [14]:
autos['price'].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

In [15]:
# Correcting data and changing dtypes for column 'odometer'.

odometer = autos['odometer']
odometer_without_km = []

for each in odometer:
    if each.endswith('km'):
        result = each.replace('km', "")
        result = result.replace(',', "")
        odometer_without_km.append(result)
    else:
        each = each.replace(",", "")
        odometer_without_km.append(each)
        
autos['odometer'] = odometer_without_km
autos['odometer'] = autos['odometer'].astype(int)

In [16]:
autos['odometer'].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer, dtype: int64

In [17]:
# Passing values to function we have created to change the name of column.

change_col_name('odometer', 'odometer_km')

autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer_km', 'registration_month', 'fuelType', 'brand',
       'unrepaired_damage', 'ad_created', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [18]:
# Lets see the what is higest and lower selling price for car
autos['price'].max()

99999999

In [19]:
autos['price'].min()

0

In [20]:
autos['price'].value_counts().sort_index(ascending=False)

99999999       1
27322222       1
12345678       3
11111111       2
10000000       1
3890000        1
1300000        1
1234566        1
999999         2
999990         1
350000         1
345000         1
299000         1
295000         1
265000         1
259000         1
250000         1
220000         1
198000         1
197000         1
194000         1
190000         1
180000         1
175000         1
169999         1
169000         1
163991         1
163500         1
155000         1
151990         1
            ... 
66             1
65             5
60             9
59             1
55             2
50            49
49             4
47             1
45             4
40             6
35             1
30             7
29             1
25             5
20             4
18             1
17             3
15             2
14             1
13             2
12             3
11             2
10             7
9              1
8              1
5              2
3              1
2             

In [21]:
# Lets see the what is higest and lower selling price for car

autos['odometer_km'].max()

150000

In [22]:
autos['odometer_km'].min()

5000

In [23]:
autos['odometer_km'].value_counts().sort_index(ascending=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

From the above observation, we can see that, price range for car starts from 0 - 99,999,999. Selling car on 0 - 99,999,999 is not correct. we will filter all cars that price range between 10,000 to 50,000, which seems more resonable.

odometer of cars seems resonable. so we change it.

In [24]:
autos.loc[autos['price'].between(10000, 50000)]

dateCrawled                                               name  \
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
37     2016-03-23 16:47:46                 Seat_Altea_XL_2.0_TDI_CR_DPF_Style   
42     2016-03-22 21:37:03  Vw_Polo_l.0_/60_PS_Blue_Motion_Technologie_Son...   
43     2016-03-08 16:51:22           Audi_A1_1.2_TFSI_Attraction_mit_Garantie   
63     2016-03-14 22:51:19                   MINI_Mini_Cooper_S___Xenon_Leder   
68     2016-04-03 17:36:59  Mini_cooper_s_clubman_/vollausstattung_/_Navi/...   
76     2016-03-22 14:52:57                         BMW_318i_neustes_Model_0Km   
90     2016-03-07 13:50:22  Audi_A3_2.0_TDI_Cabrio_S_tronic__clean_diesel_...   
95     2016-03-30 15:59:17                         Jaguar_XJ_5.0_V8_Portfolio   
98     2016-04-04 15:47:36                  Audi_A4_Cabriolet_2.0_TFSI_S_Line   
104    2016-03-12 12:57:24                Volkswagen_Multivan_T4_TDI_Highline   
105    2016-04-05 00:36:24                                      BMW_525d_Aut.   
114    2016-03-23 17:51:11  Volkswagen_Passat_Variant_2.0_TDI_BlueMotion_T...   
120    2016-03-24 22:56:04  Golf_VII_1_4_TSI_Highline_Autom_Panorama_Alcan...   
131    2016-04-03 14:49:09                                  Audi_A3_Sportback   
133    2016-03-20 02:02:00             Mercedes_Benz_ML_270_CDI_Final_Edition   
134    2016-03-10 21:37:02                           Mercedes_E200_kompressor   
138    2016-03-10 13:49:35  Ford_C_Max_Titanium_1_0L._125Ps_Benziner__Magn...   
139    2016-03-13 19:43:32                                Mercedes_Benz_E_320   
142    2016-04-01 00:46:44  Audi_A4_Avant_2.0_TDI_quattro_S_tronic__S_Line...   
153    2016-03-24 10:52:19  Porsche_Porsche_968___Turbo_S_Optik__Speed_Lin...   
155    2016-03-07 18:45:47                                  VW_Touran__140_PS   
157    2016-03-25 00:56:03  Mercedes_Benz_C_200_T_CDI_DPF_BlueEFFICIENCY_A...   
159    2016-03-22 21:56:30  Mini_Cooper_S__Leder__Xenon__Navi__Harman_Kard...   
161    2016-03-30 21:54:12                  Volkswagen_Touareg_3.0_V6_TDI_DPF   
165    2016-03-27 12:53:58           Audi_A4_2.0_TDI_S_line_Xenon_Alarmanlage   
166    2016-04-03 20:50:55                                Mercedes_Benz_200/8   
...                    ...                                                ...   
49822  2016-03-20 23:44:38  Mercedes_Benz_C_55_T_AMG_Automatik_V_MAX_offen...   
49832  2016-03-28 13:41:35  Hyundai_Santa_Fe_2.2_CRDi_/_Diesel__Automatik_...   
49833  2016-03-23 07:58:17  Mercedes_SL_320_Liebhaber_Fahrzeug___Bordcompu...   
49840  2016-03-07 20:41:52                          BMW_118d_DPF_M_Sportpaket   
49842  2016-04-02 14:42:45                 Mercedes_Benz_Vito_111_CDI_Kompakt   
49844  2016-03-26 19:43:51                                Fahrzeug_Ankauf!!!!   
49847  2016-04-03 22:58:18  Volkswagen_Golf_1.4_TSI_BlueMotion_Technology_...   
49853  2016-03-05 19:53:59                                        BMW_X3_2.0d   
49855  2016-03-24 09:36:21  Porsche_Cayman_S_PDK__Chrono_Plus__Approved__P...   
49857  2016-03-21 17:40:01  Volkswagen_Golf_1.6_TDI_BlueMotion_Technology_...   
49858  2016-03-14 14:45:47  Mercedes_Benz_Viano_3.0_CDI_lang_Automatik_Tre...   
49876  2016-03-22 17:57:24                             Audi_a5_3.0_tdi_s_line   
49893  2016-03-15 13:58:31       Mitsubishi_Pajero_3.2_DI_D_Automatik_Instyle   
49900  2016-03-26 13:56:07  Mercedes_Benz_S_350_CDI_4Matic_DPF_7G_TRONIC_B...   
49910  2016-04-03 21:39:15  Schoener_fast_neuer_Opel_Mokka_in_Zell_Mosel_m...   
49911  2016-04-02 07:54:30  BMW_X3_Euro4_TÜV_Neu_Klima_Leder_AHK_Panorama_...   
49939  2016-03-13 15:43:19                          Audi_TT_Roadster_2.0_TFSI   
49942  2016-03-09 19:43:33             Audi_A3

## Working with Date columns

In [25]:
autos[['dateCrawled','ad_created','lastSeen','registration_year']].head()

dateCrawled           ad_created             lastSeen  \
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54   
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08   
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37   
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28   
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50   

   registration_year  
0               2004  
1               1997  
2               2009  
3               2007  
4               2003

In [26]:
# Top five Crawling Date.
autos['dateCrawled'].str.split().str[0].value_counts().head()

2016-04-03    1934
2016-03-20    1891
2016-03-21    1876
2016-03-12    1839
2016-03-14    1831
Name: dateCrawled, dtype: int64

In [27]:
# Top five ad Created Date.
autos['ad_created'].str.split().str[0].value_counts().head()

2016-04-03    1946
2016-03-20    1893
2016-03-21    1886
2016-04-04    1844
2016-03-12    1831
Name: ad_created, dtype: int64

In [28]:
# Let find in which year maximum car has registerd.
autos['registration_year'].isnull().sum()

autos['registration_year'].value_counts().head()

2000    3354
2005    3015
1999    3000
2004    2737
2003    2727
Name: registration_year, dtype: int64

In Year 2000, 3354 cars has registered.


In [29]:
# Lets find how many number of car has registed in last five year.
autos['registration_year'].value_counts().sort_values(ascending=False).head()

2000    3354
2005    3015
1999    3000
2004    2737
2003    2727
Name: registration_year, dtype: int64

In [30]:
autos.loc[autos['registration_year'].between(1900, 2016)][:5]

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType  price   abtest vehicleType  registration_year    gearbox  \
0  privat   Angebot   5000  control         bus               2004    manuell   
1  privat   Angebot   8500  control   limousine               1997  automatik   
2  privat   Angebot   8990     test   limousine               2009    manuell   
3  privat   Angebot   4350  control  kleinwagen               2007  automatik   
4  privat   Angebot   1350     test       kombi               2003    manuell   

   powerPS   model  odometer_km  registration_month fuelType       brand  \
0      158  andere       150000                   3      lpg     peugeot   
1      286     7er       150000                   6   benzin         bmw   
2      102    golf        70000                   7   benzin  volkswagen   
3       71  fortwo        70000                   6   benzin       smart   
4        0   focus       150000                   7   benzin        ford   

  unrepaired_damage           ad_created  nrOfPictures  postalCode  \
0              nein  2016-03-26 00:00:00             0       79588   
1              nein  2016-04-04 00:00:00             0       71034   
2              nein  2016-03-26 00:00:00             0       35394   
3              nein  2016-03-12 00:00:00             0       33729   
4              nein  2016-04-01 00:00:00             0       39218   

              lastSeen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37  
3  2016-03-15 03:16:28  
4  2016-04-01 14:38:50

## Exploring Price by Brand

In [61]:
# Car sold by Brand
autos['brand'].value_counts(normalize=True)

volkswagen        0.21374
opel              0.10922
bmw               0.10858
mercedes_benz     0.09468
audi              0.08566
ford              0.06958
renault           0.04808
peugeot           0.02912
fiat              0.02616
seat              0.01882
skoda             0.01572
mazda             0.01514
nissan            0.01508
smart             0.01402
citroen           0.01402
toyota            0.01234
sonstige_autos    0.01092
hyundai           0.00976
volvo             0.00914
mini              0.00848
mitsubishi        0.00812
honda             0.00798
kia               0.00712
alfa_romeo        0.00658
porsche           0.00588
suzuki            0.00586
chevrolet         0.00566
chrysler          0.00362
dacia             0.00258
daihatsu          0.00256
jeep              0.00220
subaru            0.00218
land_rover        0.00198
saab              0.00160
daewoo            0.00158
trabant           0.00156
jaguar            0.00154
rover             0.00138
lancia      

In [66]:
brand_counts = autos["brand"].value_counts(normalize=True)
common_brands = brand_counts[brand_counts > .05].index
print(common_brands)

Index(['volkswagen', 'opel', 'bmw', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [69]:
brand_mean_prices = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price"].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices

{'audi': 8965,
 'bmw': 8252,
 'ford': 7105,
 'mercedes_benz': 29511,
 'opel': 5106,
 'volkswagen': 6384}

We can see that,
    - Audi, BMW, and Mercedes Benz are most expensive.
    - ford and volkswagen are mid range.
    - opel in less expensive.

## Finding the most common brand/model combinations.

In [94]:
brand_model = {}

for brand in common_brands:
    brand_model[brand] = len(autos.loc[autos['brand']==brand]['model'].unique())
    
brand_model
    

{'audi': 17,
 'bmw': 11,
 'ford': 15,
 'mercedes_benz': 19,
 'opel': 17,
 'volkswagen': 23}

Volkswagen has 23 models of cars followed by:
    - mercedes benz : 19
    - audi and opel : 17
    - ford : 15
    - bmw : 11